### PR TITLE
feat: Add Regex No Code Action Page Filter

### PR DESCRIPTION
**Pull request by victorvhs017**
## What does this PR do?

Fixes https://github.com/formbricks/internal/issues/784

#### Summary
This PR introduces regex pattern matching as a new option for page URL filtering in action classes, allowing users to create more flexible and powerful URL matching rules beyond the existing exact match, contains, starts with, and ends with options.

#### Changes Made

**Core Functionality:**
- Enhanced `testURLmatch` function to support `matchesRegex` rule type with proper error handling
- Changed return type from string ("yes"/"no") to boolean for better type safety
- Added `validateUrlFilterRegex` function for comprehensive regex pattern validation
- Implemented `createActionClassZodResolver` for form validation with regex support
- Added new action builder utilities (`buildActionObject`, `buildNoCodeAction`, `buildCodeAction`)
- Updated action class types to include `matchesRegex` rule type

**UI/UX Components:**
- **New `ActionClassInfo` component**: Displays URL filter information in a user-friendly format across multiple locations
- **New `ActionNameDescriptionFields` component**: Standardized form fields for action name and description
- **Enhanced `PageUrlSelector`**: Added regex input option with validation and error handling
- **Updated form components**: Improved error handling and validation feedback
- **Responsive design improvements**: Better layout and visual feedback for URL filter configuration

**Testing Infrastructure:**
- Comprehensive test coverage for `ActionSettingsTab` component with mocked dependencies
- Updated existing tests to use boolean return values instead of string responses
- Added test cases for regex matching scenarios and edge cases
- Enhanced test coverage for invalid regex pattern error handling
- Updated component tests for `ActionClassInfo`, `ActionNameDescriptionFields`, and form components
- Integration tests for action builder functions and validation utilities

**Docker Build Validation Improvements:**
- **Enhanced PostgreSQL Connection Verification**: Added 30-second timeout and proper error handling for database connectivity testing with detailed connection status reporting
- **Improved Health Check Polling**: Implemented robust health endpoint monitoring with 5-minute timeout (60 attempts × 5 seconds) and detailed diagnostic logging every minute
- **Container Status Monitoring**: Added continuous container status checks to detect early failures and prevent false positives
- **Better Error Diagnostics**: Enhanced logging with real-time container logs display for debugging build issues and failures
- **Network Configuration Validation**: Added network troubleshooting steps to verify port connectivity and service availability
- **PostgreSQL Extension Support**: Added initialization steps for required PostgreSQL extensions during build validation
- **Timeout Handling**: Implemented smart timeout functions in docker start scripts with Alpine Linux coreutils fallback support
- **Host Network Integration**: Added support for `host.docker.internal` networking on Linux for better container-to-host communication
- **Progress Monitoring**: Added detailed progress reporting during health checks with elapsed time tracking
- **Graceful Cleanup**: Improved container cleanup process with proper error handling and resource management

**Internationalization:**
- Added translations for regex-related UI elements across all supported languages:
  - English (en-US), German (de-DE), French (fr-FR)
  - Portuguese Brazil (pt-BR), Portuguese Portugal (pt-PT), Italian (it-IT)
- Comprehensive error message translations for:
  - Invalid regex pattern errors
  - Invalid match type scenarios
  - Form validation messages
- Translation keys for new UI components and help text

**Form Validation & Error Handling:**
- Robust regex pattern validation with user-friendly error messages
- Graceful handling of malformed regex patterns
- Localized error messages for better user experience
- Enhanced form validation using Zod resolvers
- Real-time validation feedback for URL pattern input

#### Technical Details
- **New Rule Type:** `matchesRegex` - allows users to input JavaScript regex patterns for URL matching
- **Error Handling:** Comprehensive error catching for invalid regex with localized error messages
- **Type Safety:** Enhanced TypeScript types for better development experience
- **Backward Compatibility:** All existing URL matching rules (exactMatch, contains, startsWith, endsWith) remain unchanged
- **Performance:** Optimized regex compilation and testing for better performance
- **Architecture:** Modular component design for better maintainability and reusability
- **CI/CD Reliability:** Significantly improved Docker build validation workflow with better error detection and debugging capabilities

#### UI/UX Improvements
- **New regex input option** in page filter configuration with helpful placeholder text
- **Enhanced visual feedback** when surveys would/wouldn't be shown based on URL rules
- **Improved component organization** with reusable `ActionClassInfo` component
- **Standardized form fields** with consistent styling and validation
- **Better error messaging** with clear, actionable feedback for users
- **Responsive design** improvements for mobile and desktop views

#### Code Quality Improvements
- **Modular architecture**: Separated concerns with dedicated utility functions and components
- **Enhanced testing**: Comprehensive test coverage with proper mocking and edge case handling
- **Type safety**: Improved TypeScript definitions and validation schemas
- **Code reusability**: New shared components reduce code duplication
- **Maintainability**: Better organized codebase with clear separation of responsibilities
- **CI/CD Robustness**: More reliable build validation with comprehensive error handling and diagnostics

Other UI updates:

Before:
image

After:
image

Before:
image

After:
image

Before:
image

After:
image

Before:
image

After:
image

---

#### **Basic Regex Functionality**
- [ ] Create a new action or edit an existing one
- [ ] In the page filter section, verify "Matches regex" appears as an option
- [ ] Test basic regex patterns:
  - [ ] `user/\d+` should match URLs like `/user/123` but not `/user/abc`
  - [ ] `dashboard$` should match URLs ending with "dashboard"
  - [ ] `^https://app` should match URLs starting with "https://app"

#### **Error Handling**
- [ ] Test invalid regex patterns:
  - [ ] Enter `[invalid-regex` (unclosed bracket)
  - [ ] Enter `*invalid` (invalid quantifier)
  - [ ] Verify appropriate error messages appear in the UI
  - [ ] Test error messages are properly localized

#### **Component Integration**
- [ ] Verify `ActionClassInfo` component displays URL filters correctly
- [ ] Test `ActionNameDescriptionFields` component functionality
- [ ] Ensure form validation works as expected with new Zod resolvers
- [ ] Test responsive design across different screen sizes

#### **Docker Build Validation**
- [ ] Verify Docker build validation workflow passes with improved error handling
- [ ] Test PostgreSQL connection timeout and error reporting
- [ ] Confirm health check polling works correctly with detailed logging
- [ ] Validate container status monitoring detects failures properly
- [ ] Test network configuration validation steps

#### **Integration Testing**
- [ ] Create an action with regex page filter
- [ ] Test the action triggers correctly on matching pages
- [ ] Test the action doesn't trigger on non-matching pages
- [ ] Verify existing actions with other filter types still work correctly
- [ ] Test backward compatibility with existing action configurations

#### **Internationalization Testing**
- [ ] Switch between different languages and verify all new text elements are translated
- [ ] Test error messages appear in the correct language
- [ ] Verify translation keys are working properly with Tolgee integration 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: apps/web/lib/utils/url.ts [hunk 1]
```diff
@@ -3,23 +3,34 @@ import { TActionClassPageUrlRule } from "@formbricks/types/action-classes";
...
-      return !testUrl.includes(pageUrlValue) ? "yes" : "no";
+      return !testUrl.includes(pageUrlValue);
+    case "matchesRegex":
+      try {
+        regex = new RegExp(pageUrlValue);
+      } catch {
+        throw new Error(t("environments.actions.invalid_regex"));
+      }
+
+      return regex.test(testUrl);
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply apps/web/lib/utils/url.ts:31]: That is intentional. We fail closed in the SDK if an invalid regex pattern is saved in the database somehow.

File: packages/js-core/src/lib/common/utils.ts [hunk 2]
```diff
@@ -205,6 +205,8 @@ export const checkUrlMatch = (
   pageUrlValue: string,
   pageUrlRule: TActionClassPageUrlRule
 ): boolean => {
+  let regex: RegExp;
+
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply packages/js-core/src/lib/common/utils.ts:209]: SonarQube directive goes the other way and will complain if we scope it inside the case.

File: apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx [hunk 3]
```diff
@@ -202,21 +208,29 @@ const UrlInput = ({
...
-                    {...field}
-                    placeholder="e.g. https://app.com/dashboard"
-                    autoComplete="off"
-                    isInvalid={!!error?.message}
-                  />
-                </FormControl>
-              </FormItem>
-            )}
+            render={({ field, fieldState: { error } }) => {
+              const ruleValue = ruleValues?.[index]?.rule;
```
Comment (Dhruwang) [inline apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx]: `ruleValues[index]?.rule` would work fine here

File: apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx [hunk 4]
```diff
@@ -193,6 +198,7 @@ const UrlInput = ({
                       <SelectItem value="notContains">
                         {t("environments.actions.does_not_contain")}
                       </SelectItem>
+                      <SelectItem value="matchesRegex">{t("environments.actions.matches_regex")}</SelectItem>
```
Comment (Dhruwang) [inline apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx]: Can we refactor this to use mapping instead ?

File: apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx [hunk 5]
```diff
@@ -34,29 +35,35 @@ interface PageUrlSelectorProps {
...
-    if (isMatch === "no") toast.error("Your survey would not be shown.");
+      setIsMatch(match);
+      if (match) toast.success(t("environments.actions.your_survey_would_be_shown_on_this_url"));
+      if (!match) toast.error(t("environments.actions.your_survey_would_not_be_shown"));
+    } catch (error) {
+      toast.error(error.message);
+    }
   };
 
+  const isMatchClass = useCallback(() => {
```
Comment (Dhruwang) [inline apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx]: Since we're using the return value directly, consider renaming to `matchClass` and switching from `useCallback` to `useMemo`.

File: apps/web/modules/ui/components/action-class-info/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,375 @@
+import "@testing-library/jest-dom/vitest";
+import { cleanup, render, screen } from "@testing-library/react";
+import { afterEach, describe, expect, test, vi } from "vitest";
+import { TActionClass } from "@formbricks/types/action-classes";
+import { ActionClassInfo } from "./index";
+
+// Mock the tolgee translation
+vi.mock("@tolgee/react", () => ({
```
Comment (Dhruwang) [inline apps/web/modules/ui/components/action-class-info/index.test.tsx]: We already have this in global mocks

**Review by Dhruwang — CHANGES_REQUESTED** [review]
Looks good ! Functionality works well 🚀

Here are some issues that i noticed. Some issue are not related to this PR but since we are already tweaking things here, we can fix those inconsistencies 😊

1. I noticed some container with color `bg-slate-50` in the background here

Screenshot 2025-07-25 at 4 37 58 PM

2. Styling inconsistencies

Screenshot 2025-07-25 at 4 40 31 PM

3. No check for validity of regex when trying to save the action


https://github.com/user-attachments/assets/f6552b92-5391-4c67-a438-c2ca6fc9c4dd

File: apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx [hunk 7]
```diff
@@ -34,29 +60,35 @@ interface PageUrlSelectorProps {
...
 
-    setIsMatch(isMatch);
-    if (isMatch === "yes") toast.success("Your survey would be shown on this URL.");
-    if (isMatch === "no") toast.error("Your survey would not be shown.");
+      setIsMatch(match);
+      if (match) toast.success(t("environments.actions.your_survey_would_be_shown_on_this_url"));
+      if (!match) toast.error(t("environments.actions.your_survey_would_not_be_shown"));
+    } catch (error) {
+      toast.error(error.message);
+    }
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx:84]: It can only throw in the testURLMatch, and it always returns an error.

File: apps/web/modules/survey/editor/lib/action-builder.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { TActionClassInput } from "@formbricks/types/action-classes";
+
+export const buildActionObject = (data: TActionClassInput, environmentId: string) => {
+  if (data.type === "noCode") {
+    return buildNoCodeAction(data, environmentId);
+  }
+  return buildCodeAction(data, environmentId);
+};
+
+export const buildNoCodeAction = (data: TActionClassInput, environmentId: string) => {
```
Comment (Dhruwang) [inline apps/web/modules/survey/editor/lib/action-builder.ts]: we can remove the export here as this function is only being used in this file

File: apps/web/modules/survey/editor/lib/action-builder.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+          innerHtml: data.noCodeConfig.elementSelector.innerHtml,
+        },
+      },
+    };
+  }
+
+  return baseAction;
+};
+
+export const buildCodeAction = (data: TActionClassInput, environmentId: string) => {
```
Comment (Dhruwang) [inline apps/web/modules/survey/editor/lib/action-builder.ts]: we can remove the export here as this function is only being used in this file

File: apps/web/modules/survey/editor/lib/action-builder.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+export const buildActionObject = (data: TActionClassInput, environmentId: string) => {
+  if (data.type === "noCode") {
+    return buildNoCodeAction(data, environmentId);
+  }
+  return buildCodeAction(data, environmentId);
+};
+
+export const buildNoCodeAction = (data: TActionClassInput, environmentId: string) => {
+  if (data.type !== "noCode") {
```
Comment (Dhruwang) [inline apps/web/modules/survey/editor/lib/action-builder.ts:12]: Duplicate check. We already do this check in `buildActionObject`

File: apps/web/modules/survey/editor/lib/action-validation.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+};
+
+export const validateActionData = (
+  data: TActionClassInput,
+  isReadOnly: boolean,
+  actionClassNames: string[],
+  actionClassKeys: string[],
+  t: any
```
Comment (Dhruwang) [inline apps/web/modules/survey/editor/lib/action-validation.ts]: instead of any we can use `TFnType`

File: apps/web/modules/survey/editor/lib/action-validation.ts
Comment (Dhruwang) [inline apps/web/modules/survey/editor/lib/action-validation.ts:1]: We can remove exports for validatePermissions, validateActionNames, validateActionKey ,validateCssSelector and validateRegexPatterns

File: apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx [hunk 12]
```diff
@@ -16,16 +16,42 @@ import {
...
 } from "react-hook-form";
 import toast from "react-hot-toast";
-import { TActionClassInput, TActionClassPageUrlRule } from "@formbricks/types/action-classes";
+import {
+  ACTION_CLASS_PAGE_URL_RULES,
+  TActionClassInput,
+  TActionClassPageUrlRule,
+} from "@formbricks/types/action-classes";
+
+const getRuleLabel = (rule: TActionClassPageUrlRule, t: (key: string) => string): string => {
```
Comment (Dhruwang) [inline apps/web/modules/ui/components/no-code-action-form/components/page-url-selector.tsx]: Use type `TFnType` for type of `t` or tolgee will considered those keys as unused

**Review by Dhruwang — CHANGES_REQUESTED** [review]
I came across an edge case where, if a code action is created using the 'Exactly match' operator with an invalid regex string, switching it to the 'Matches Regex' operator doesn’t trigger any validation error.


https://github.com/user-attachments/assets/8555d9dd-e146-414f-8973-929a11782047

File: apps/web/modules/survey/editor/lib/action-builder.ts (hunk 8 above)
Comment (victorvhs017) [reply apps/web/modules/survey/editor/lib/action-builder.ts]: We export it to test it.

File: apps/web/modules/survey/editor/lib/action-builder.ts (hunk 9 above)
Comment (victorvhs017) [reply apps/web/modules/survey/editor/lib/action-builder.ts]: We export it to test it.

File: apps/web/modules/survey/editor/lib/action-validation.ts
Comment (victorvhs017) [reply apps/web/modules/survey/editor/lib/action-validation.ts:1]: Same here, we export them to better test the functions.

File: apps/web/modules/survey/editor/lib/action-builder.ts (hunk 10 above)
Comment (victorvhs017) [reply apps/web/modules/survey/editor/lib/action-builder.ts:12]: This works more as a type guard here. If we remove, we'll have to do some type assertions, so I think this is a more appropriate way of doing it.

But we should be returning a Tolgee translation string, I'll update that!

**Comment by victorvhs017** [comment]
> I came across an edge case where, if a code action is created using the 'Exactly match' operator with an invalid regex string, switching it to the 'Matches Regex' operator doesn’t trigger any validation error.
> 
>  Screen-Recording-2025-07-29-at-11.12.19.AM.mp4

I forgot about the Actions tab on the side menu. I was testing only inside the survey editor.

When I checked the Actions tab menu, I saw a lot of duplicate code, which is why I did a large refactoring in the latest commit.

File: apps/web/modules/ui/components/code-action-form/index.test.tsx [hunk 13]
```diff
@@ -31,6 +31,7 @@ vi.mock("@/modules/ui/components/form", () => ({
   },
   FormItem: ({ children }: { children: React.ReactNode }) => <div data-testid="form-item">{children}</div>,
   FormLabel: ({ children }: { children: React.ReactNode }) => <div data-testid="form-label">{children}</div>,
+  FormError: () => <div data-testid="form-error">Form Error</div>,
 }));
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply apps/web/modules/ui/components/code-action-form/index.test.tsx:35]: Not relevant to the test.

**Review by Dhruwang — DISMISSED** [review]

File: .github/workflows/docker-build-validation.yml [hunk 14]
```diff
@@ -59,18 +59,34 @@
...
+            
+          else
+            echo "❌ PostgreSQL connection failed after 30 seconds"
+            exit 1
+          fi
 
           # Show network configuration
           echo "Network configuration:"
-          ip addr show
           netstat -tulpn | grep 5432 || echo "No process listening on port 5432"
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply .github/workflows/docker-build-validation.yml:88]: It was available in the test check

File: .github/workflows/docker-build-validation.yml [hunk 15]
```diff
@@ -89,50 +105,79 @@
...
-          # Try connecting to PostgreSQL from inside the container
+          # Try connecting to PostgreSQL from inside the container (with correct package manager)
           echo "Testing PostgreSQL connection from inside container..."
-          docker exec formbricks-test sh -c 'apt-get update && apt-get install -y postgresql-client && PGPASSWORD=test psql -h host.docker.internal -U test -d formbricks -c "\dt" || echo "Failed to connect to PostgreSQL from container"'
+          docker exec formbricks-test sh -c '
+            echo "Installing postgresql-client using apk..."
+            apk add --no-cache postgresql-client
+            echo "Testing connection..."
+            timeout 10 sh -c "PGPASSWORD=test psql -h host.docker.internal -U test -d formbricks -c \"SELECT 1;\"" || echo "❌ Failed to connect to PostgreSQL from container"
+          '
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply .github/workflows/docker-build-validation.yml]: It worked in the test run

File: .github/workflows/docker-build-validation.yml [hunk 16]
```diff
@@ -59,18 +59,32 @@ jobs:
             database_url=${{ secrets.DUMMY_DATABASE_URL }}
             encryption_key=${{ secrets.DUMMY_ENCRYPTION_KEY }}
 
-      - name: Verify PostgreSQL Connection
+      - name: Verify and Initialize PostgreSQL
         run: |
           echo "Verifying PostgreSQL connection..."
           # Install PostgreSQL client to test connection
           sudo apt-get update && sudo apt-get install -y postgresql-client
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply .github/workflows/docker-build-validation.yml:67]: It's working fine in the tests

File: .github/workflows/docker-build-validation.yml [hunk 17]
```diff
@@ -89,65 +103,42 @@
...
+              break
             fi
             
-            echo "Waiting 15 seconds before next attempt..."
-            sleep 15
+            # Wait 5 seconds before next attempt
+            sleep 5
           done
 
           # Show full container logs for debugging
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply .github/workflows/docker-build-validation.yml]: Not needed right now, the check is running correctly.

File: apps/web/scripts/docker/next-start.sh [hunk 18]
```diff
@@ -2,12 +2,64 @@
 
 set -eu
 export NODE_ENV=production
+
+# Function to run command with timeout if available, or without timeout as fallback
+run_with_timeout() {
+  _timeout_duration="$1"
+  _timeout_name="$2"
+  shift 2
+
```
[the remark this answers is not in the document]
Comment (victorvhs017) [reply apps/web/scripts/docker/next-start.sh:11]: not relevant for action runs
